### ISSUE TITLE
Object rest, side effect order and function scope

### ISSUE THREAD

**Issue by magic-akari** [evanw/esbuild#3749]
# Playground
https://esbuild.github.io/try/#dAAwLjIwLjIALS10YXJnZXQ9Y2hyb21lNTkAZnVuY3Rpb24gYSgpIHsKICBjb25zb2xlLmxvZygiYSIpOwogIHJldHVybiAiYSI7Cn0KCmZ1bmN0aW9uIGIoKSB7CiAgY29uc29sZS5sb2coImIiKTsKICByZXR1cm4gImIiOwp9CgpmdW5jdGlvbiBjKCkgewogIGNvbnNvbGUubG9nKCJjIik7CiAgcmV0dXJuICJjIjsKfQoKZnVuY3Rpb24gZm9vKHsgYWEsIC4uLmJiIH0sIGNjID0gYWEoKSkgewogIGNvbnNvbGUubG9nKGFhLCBiYiwgY2MpCn0KCgoKZm9vKHsgYWE6IGEsIGJiOiBiLCBjYzogYyB9KQ

# Input

```JavaScript
function a() {
  console.log("a");
  return "a";
}

function b() {
  console.log("b");
  return "b";
}

function c() {
  console.log("c");
  return "c";
}

function foo({ aa, ...bb }, cc = aa()) {
  console.log(aa, bb, cc)
}



foo({ aa: a, bb: b, cc: c })
```

with `--target=chrome59`

# Expected Result
print some function and objects.

# Actual Result
Uncaught ReferenceError: aa is not defined

# Relates issue
https://github.com/babel/babel/issues/16470

**Comment by evanw** [comment]
This comment is relevant:

https://github.com/evanw/esbuild/blob/51f67d4523bdea1e224753078b03dc12ea0ec70b/internal/js_parser/js_parser_lower.go#L284-L313

**Comment by magic-akari** [comment]
esbuild considers scope to be more important than side effect order.
However, in this example, the reference to `aa` is still incorrect.

**Comment by magic-akari** [comment]
Thank you for the clear comments in esbuild.
This issue was actually discovered when I was reading the esbuild source code.
However, I feel that it might not be worth solving, or it might be unsolvable. I'm just bringing it up here for the record.

**Comment by evanw** [comment]
Yes, makes sense. Thanks for the report. When I get the time to think about this, I'll have to load this back in my head to figure out what to do about it, if anything. That comment (and any relevant surrounding commit logs from then) will be a good starting point for me.
